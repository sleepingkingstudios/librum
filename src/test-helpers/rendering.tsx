import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
