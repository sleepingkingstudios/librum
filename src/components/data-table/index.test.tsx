import * as React from 'react';

import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { DataTable } from './index';
import type {
  DataTableColumn,
  DataMappingFunction,
} from './types';

type SpaceMission = {
  deltaV: string,
  launchDate?: string,
  missionName?: string,
  missionObjective: string,
  ordinal: string,
  program: string,
  programId?: number,
};

type SpaceProgram = {
  id: number,
  name: string,
};

describe('<DataTable />', () => {
  const Actions = (): JSX.Element => (
    <span className="data-table-actions">Launch</span>
  );
  const columns: DataTableColumn[] = [
    {
      name: 'missionName',
      value: ({ data }: { data: SpaceMission }) => (
        `${data.program} ${data.ordinal}`
      ),
    },
    {
      label: 'Objective',
      name: 'missionObjective',
    },
    {
      name: 'deltaV',
      numeric: true,
    },
    {
      align: 'center',
      default: 'TBD',
      name: 'launchDate',
    },
    {
      label: false,
      name: 'actions',
      value: Actions,
    },
  ];
  const name = 'spaceMissions';
  const expectedHeader = [
    'Mission Name',
    'Objective',
    'Delta V',
    'Launch Date',
    'Â ',
  ];

  describe('when there is no matching data', () => {
    const data = {};
    const expectedMessage = 'There are no matching space missions.';

    it('should render the header cells', () => {
      const { getAllByRole } = render(
        <DataTable columns={columns} data={data} name={name} />,
      );
      const cells = getAllByRole('columnheader');

      expect(cells).toHaveLength(5);
      expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
    });

    it('should display the empty message', () => {
      const { getByRole } = render(
        <DataTable columns={columns} data={data} name={name} />,
      );
      const cell = getByRole('cell');

      expect(cell.textContent).toBe(expectedMessage);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTable columns={columns} data={data} name={name} />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there are no rows', () => {
    const data: { 'spaceMissions': SpaceMission[] } = { 'spaceMissions': [] };
    const expectedMessage = 'There are no matching space missions.';

    it('should render the header cells', () => {
      const { getAllByRole } = render(
        <DataTable columns={columns} data={data} name={name} />,
      );
      const cells = getAllByRole('columnheader');

      expect(cells).toHaveLength(5);
      expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
    });

    it('should display the empty message', () => {
      const { getByRole } = render(
        <DataTable columns={columns} data={data} name={name} />,
      );
      const cell = getByRole('cell');

      expect(cell.textContent).toBe(expectedMessage);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTable columns={columns} data={data} name={name} />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there are many rows', () => {
    const data: { spaceMissions: SpaceMission[] } = {
      spaceMissions: [
        {
          deltaV: '1,000 km/s',
          launchDate: '2050',
          missionObjective: 'Titan',
          ordinal: 'IX',
          program: 'Chronos',
        },
        {
          deltaV: '10,000 km/s',
          launchDate: '2150',
          missionObjective: 'Alpha Centauri A',
          ordinal: 'X',
          program: 'Chiron',
        },
        {
          deltaV: '100,000 km/s',
          launchDate: '3050',
          missionObjective: 'Andromeda',
          ordinal: 'XI',
          program: 'Perseus',
        },
      ],
    };
    const expectedBody = [
      'Chronos IX',
      'Titan',
      '1,000 km/s',
      '2050',
      'Launch',
      'Chiron X',
      'Alpha Centauri A',
      '10,000 km/s',
      '2150',
      'Launch',
      'Perseus XI',
      'Andromeda',
      '100,000 km/s',
      '3050',
      'Launch',
    ];

    it('should render the header cells', () => {
      const { getAllByRole } = render(
        <DataTable columns={columns} data={data} name={name} />,
      );
      const cells = getAllByRole('columnheader');

      expect(cells).toHaveLength(5);
      expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
    });

    it('should display the formatted data', () => {
      const { getAllByRole } = render(
        <DataTable columns={columns} data={data} name={name} />,
      );
      const cells = getAllByRole('cell');

      expect(cells).toHaveLength(15);
      expect(cells.map(cell => cell.textContent)).toEqual(expectedBody);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTable columns={columns} data={data} name={name} />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with collapse: true', () => {
    describe('when there are no rows', () => {
      const data: { 'spaceMissions': SpaceMission[] } = { 'spaceMissions': [] };
      const expectedMessage = 'There are no matching space missions.';

      it('should render the header cells', () => {
        const { getAllByRole } = render(
          <DataTable collapse columns={columns} data={data} name={name} />,
        );
        const cells = getAllByRole('columnheader');

        expect(cells).toHaveLength(5);
        expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
      });

      it('should display the empty message', () => {
        const { getByRole } = render(
          <DataTable collapse columns={columns} data={data} name={name} />,
        );
        const cell = getByRole('cell');

        expect(cell.textContent).toBe(expectedMessage);
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTable collapse columns={columns} data={data} name={name} />,
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });


    describe('when there are many rows', () => {
      const data: { spaceMissions: SpaceMission[] } = {
        spaceMissions: [
          {
            deltaV: '1,000 km/s',
            launchDate: '2050',
            missionObjective: 'Titan',
            ordinal: 'IX',
            program: 'Chronos',
          },
          {
            deltaV: '10,000 km/s',
            launchDate: '2150',
            missionObjective: 'Alpha Centauri A',
            ordinal: 'X',
            program: 'Chiron',
          },
          {
            deltaV: '100,000 km/s',
            launchDate: '3050',
            missionObjective: 'Andromeda',
            ordinal: 'XI',
            program: 'Perseus',
          },
        ],
      };
      const expectedBody = [
        'Chronos IX',
        'Titan',
        '1,000 km/s',
        '2050',
        'Launch',
        'Chiron X',
        'Alpha Centauri A',
        '10,000 km/s',
        '2150',
        'Launch',
        'Perseus XI',
        'Andromeda',
        '100,000 km/s',
        '3050',
        'Launch',
      ];

      it('should render the header cells', () => {
        const { getAllByRole } = render(
          <DataTable collapse columns={columns} data={data} name={name} />,
        );
        const cells = getAllByRole('columnheader');

        expect(cells).toHaveLength(5);
        expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
      });

      it('should display the formatted data', () => {
        const { getAllByRole } = render(
          <DataTable collapse columns={columns} data={data} name={name} />,
        );
        const cells = getAllByRole('cell');

        expect(cells).toHaveLength(15);
        expect(cells.map(cell => cell.textContent)).toEqual(expectedBody);
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTable collapse columns={columns} data={data} name={name} />,
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with emptyMessage: component', () => {
    const EmptyMessage = (): JSX.Element => (
      <span className="data-table-empty-message">
        You are not going to space today.
      </span>
    );

    describe('when there are no rows', () => {
      const data: { 'spaceMissions': SpaceMission[] } = { 'spaceMissions': [] };
      const expectedMessage = 'You are not going to space today.';

      it('should render the header cells', () => {
        const { getAllByRole } = render(
          <DataTable
            columns={columns}
            data={data}
            emptyMessage={EmptyMessage}
            name={name}
          />,
        );
        const cells = getAllByRole('columnheader');

        expect(cells).toHaveLength(5);
        expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
      });

      it('should display the empty message', () => {
        const { getByRole } = render(
          <DataTable
            columns={columns}
            data={data}
            emptyMessage={EmptyMessage}
            name={name}
          />,
        );
        const cell = getByRole('cell');

        expect(cell.textContent).toBe(expectedMessage);
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTable
            columns={columns}
            data={data}
            emptyMessage={EmptyMessage}
            name={name}
          />,
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with emptyMessage: string', () => {
    const emptyMessage = 'You are not going to space today.';

    describe('when there are no rows', () => {
      const data: { 'spaceMissions': SpaceMission[] } = { 'spaceMissions': [] };

      it('should render the header cells', () => {
        const { getAllByRole } = render(
          <DataTable
            columns={columns}
            data={data}
            emptyMessage={emptyMessage}
            name={name}
          />,
        );
        const cells = getAllByRole('columnheader');

        expect(cells).toHaveLength(5);
        expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
      });

      it('should display the empty message', () => {
        const { getByRole } = render(
          <DataTable
            columns={columns}
            data={data}
            emptyMessage={emptyMessage}
            name={name}
          />,
        );
        const cell = getByRole('cell');

        expect(cell.textContent).toBe(emptyMessage);
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTable
            columns={columns}
            data={data}
            emptyMessage={emptyMessage}
            name={name}
          />,
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with mapping: function', () => {
    type SpaceData = {
      spaceMissions: SpaceMission[],
      spacePrograms: SpaceProgram[],
    };

    const mapping: DataMappingFunction = (data: SpaceData) => {
      const {
        spaceMissions,
        spacePrograms,
      } = data;

      return spaceMissions.map(
        (mission: SpaceMission) => {
          const match = spacePrograms.find(
            (program) => (program.id === mission.programId)
          );
          const programName = match ? match.name : 'Unknown';

          return {
            ...mission,
            program: programName,
          };
        }
      );
    };
    const data = {
      spaceMissions: [
        {
          deltaV: '1,000 km/s',
          launchDate: '2050',
          missionObjective: 'Titan',
          ordinal: 'IX',
          programId: 0,
        },
        {
          deltaV: '10,000 km/s',
          launchDate: '2150',
          missionObjective: 'Alpha Centauri A',
          ordinal: 'X',
          programId: 1,
        },
        {
          deltaV: '100,000 km/s',
          launchDate: '3050',
          missionObjective: 'Andromeda',
          ordinal: 'XI',
          programId: 2,
        },
      ],
      spacePrograms: [
        {
          id: 0,
          name: 'Chronos',
        },
        {
          id: 1,
          name: 'Chiron',
        },
        {
          id: 2,
          name: 'Perseus'
        },
      ],
    };
    const expectedBody = [
      'Chronos IX',
      'Titan',
      '1,000 km/s',
      '2050',
      'Launch',
      'Chiron X',
      'Alpha Centauri A',
      '10,000 km/s',
      '2150',
      'Launch',
      'Perseus XI',
      'Andromeda',
      '100,000 km/s',
      '3050',
      'Launch',
    ];

    it('should render the header cells', () => {
      const { getAllByRole } = render(
        <DataTable
          columns={columns}
          data={data}
          mapping={mapping}
          name={name}
        />,
      );
      const cells = getAllByRole('columnheader');

      expect(cells).toHaveLength(5);
      expect(cells.map(cell => cell.textContent)).toEqual(expectedHeader);
    });

    it('should display the formatted data', () => {
      const { getAllByRole } = render(
        <DataTable
          columns={columns}
          data={data}
          mapping={mapping}
          name={name}
        />,
      );
      const cells = getAllByRole('cell');

      expect(cells).toHaveLength(15);
      expect(cells.map(cell => cell.textContent)).toEqual(expectedBody);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTable
          columns={columns}
          data={data}
          mapping={mapping}
          name={name}
        />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with scope: value', () => {
    const scope = 'rocketry';

    describe('when there are no rows', () => {
      const data: { 'spaceMissions': SpaceMission[] } = { 'spaceMissions': [] };

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTable columns={columns} data={data} name={name} scope={scope} />,
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when there are many rows', () => {
      const data: { spaceMissions: SpaceMission[] } = {
        spaceMissions: [
          {
            deltaV: '1,000 km/s',
            launchDate: '2050',
            missionObjective: 'Titan',
            ordinal: 'IX',
            program: 'Chronos',
          },
          {
            deltaV: '10,000 km/s',
            launchDate: '2150',
            missionObjective: 'Alpha Centauri A',
            ordinal: 'X',
            program: 'Chiron',
          },
          {
            deltaV: '100,000 km/s',
            launchDate: '3050',
            missionObjective: 'Andromeda',
            ordinal: 'XI',
            program: 'Perseus',
          },
        ],
      };

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTable columns={columns} data={data} name={name} scope={scope} />,
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });
});
