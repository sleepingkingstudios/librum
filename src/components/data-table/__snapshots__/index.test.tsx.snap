// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataTable /> when there are many rows should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Chronos IX
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Titan
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            1,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            2050
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Chiron X
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Alpha Centauri A
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            10,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            2150
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Perseus XI
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Andromeda
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            100,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            3050
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<DataTable /> when there are no rows should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-empty"
            colspan="5"
          >
            There are no matching space missions.
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<DataTable /> when there is no matching data should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-empty"
            colspan="5"
          >
            There are no matching space missions.
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<DataTable /> with collapse: true when there are many rows should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-collapse data-table-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Chronos IX
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Titan
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            1,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            2050
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Chiron X
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Alpha Centauri A
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            10,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            2150
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Perseus XI
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Andromeda
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            100,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            3050
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<DataTable /> with collapse: true when there are no rows should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-collapse data-table-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-empty"
            colspan="5"
          >
            There are no matching space missions.
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<DataTable /> with emptyMessage: component when there are no rows should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-empty"
            colspan="5"
          >
            <span
              class="data-table-empty-message"
            >
              You are not going to space today.
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<DataTable /> with emptyMessage: string when there are no rows should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-empty"
            colspan="5"
          >
            You are not going to space today.
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<DataTable /> with mapping: function should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Chronos IX
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Titan
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            1,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            2050
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Chiron X
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Alpha Centauri A
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            10,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            2150
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Perseus XI
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Andromeda
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            100,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            3050
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<DataTable /> with scope: value when there are many rows should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-rocketry-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Chronos IX
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Titan
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            1,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            2050
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Chiron X
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Alpha Centauri A
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            10,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            2150
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
        <tr>
          <td
            class="data-table-cell-mission-name"
            data-label="Mission Name"
          >
            Perseus XI
          </td>
          <td
            class="data-table-cell-mission-objective"
            data-label="Objective"
          >
            Andromeda
          </td>
          <td
            class="data-table-cell-delta-v text-right lining-nums tabular-nums"
            data-label="Delta V"
          >
            100,000 km/s
          </td>
          <td
            class="data-table-cell-launch-date text-center"
            data-label="Launch Date"
          >
            3050
          </td>
          <td
            class="data-table-cell-actions"
            data-label="Actions"
          >
            <span
              class="data-table-actions"
            >
              Launch
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;

exports[`<DataTable /> with scope: value when there are no rows should match the snapshot 1`] = `
<DocumentFragment>
  <figure
    class="data-table data-table-rocketry-space-missions"
  >
    <table>
      <thead
        class="data-table-header"
      >
        <tr>
          <th
            class="data-table-header-mission-name"
          >
            Mission Name
          </th>
          <th
            class="data-table-header-mission-objective"
          >
            Objective
          </th>
          <th
            class="data-table-header-delta-v text-right"
          >
            Delta V
          </th>
          <th
            class="data-table-header-launch-date text-center"
          >
            Launch Date
          </th>
          <th
            class="data-table-header-actions"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="data-table-empty"
            colspan="5"
          >
            There are no matching space missions.
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</DocumentFragment>
`;
